packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/MdbJavaEditorToolbar.kt

```diff
@@ -1,164 +1,181 @@
+            coroutineScope.launchChildBackground {
+                dataSourceComboBox.connecting = true
+                val connectionManager = DatabaseConnectionManager.getInstance()
+                val connectionHandler =
+                    connectionManager
+                        .build(project, dataSource)
+                        .setRequestor(ConnectionRequestor.Anonymous())
+                        .setAskPassword(true)
+                        .setRunConfiguration(
+                            ConsoleRunConfiguration.newConfiguration(project).apply {
+                                setOptionsFromDataSource(dataSource)
+                            },
+                        )
+
+                val connectionJob = runCatching { connectionHandler.create()?.get() }
+                connectionJob.onFailure {
+                    dataSourceComboBox.connecting = false
+                    dataSourceComboBox.failedConnection = dataSource
+                    reloadDatabases()
+                    return@launchChildBackground
+                }
+
+                val connection = connectionJob.getOrNull()
+                dataSourceComboBox.connecting = false
+
+                // could not connect, do nothing
+                if (connection == null || !dataSource.isConnected()) {
+                    dataSourceComboBox.selectedDataSource = null // remove data source because we didn't connect
+                    reloadDatabases()
+                    return@launchChildBackground
+                }
+
+                onConnected(dataSource)
+                reloadDatabases()
             }
+        } else {
+            onConnected(dataSource)
+            reloadDatabases()
         }
     }
 
-    private fun selectDataSourceWithId(id: String?) {
-        id?.let {
-            val dataSourceIndex =
-                (connectionComboBox.model as DefaultComboBoxModel<LocalDataSource?>)
-                    .asSequence()
-                    .toList()
-                    .indexOf { it?.uniqueId == id }
-            if (dataSourceIndex == -1) {
-                connectionComboBox.selectedItem = null
-            } else {
-                connectionComboBox.selectedIndex = dataSourceIndex
-            }
-        } ?: run {
-            connectionComboBox.selectedItem = null
+    internal fun onDataSourceUnselected() {
+        onDisconnected()
+        reloadDatabases()
+    }
+
+    fun reloadDataSources(dataSources: List<LocalDataSource>) {
+        dataSourceComboBox.dataSources = dataSources.filter { it.isMongoDbDataSource() }
+    }
+
+    fun reloadDatabases() {
+        dataSourceComboBox.selectedDataSource?.let {
+val readModel = project.getService(DataGripBasedReadModelProvider::class.java)
+val databases = readModel.slice(dataSourceComboBox.selectedDataSource!!, ListDatabases.Slice)
+databaseComboBox.databases = databases.databases.map { it.name }
+} ?: run {
+databaseComboBox.databases = emptyList()
```

Comment:
I think we need to check our Spotless config. I noticed that when we don't apply the formatting ourselves, it screws up on the indentation.

Reply:
That sounds good, I'll create a ticket for this and we can spend some time at some point. For now I'll reformat the file and push the changes. Maybe a good change is to replace the gitHook to do spotlessCheck instead of spotlessApply.